exporter/awsxrayexporter/translator/segment.go

```diff
@@ -0,0 +1,320 @@
+//    1-58406520-a006649127e371903a2de979. This includes:
+//  * The version number, that is, 1.
+//  * The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+//  * For example, 10:00AM December 2nd, 2016 PST in epoch time is 1480615200 seconds,
+//    or 58406520 in hexadecimal.
+//  * A 96-bit identifier for the trace, globally unique, in 24 hexadecimal digits.
+func convertToAmazonTraceID(traceID []byte) string {
+	const (
+		// maxAge of 28 days.  AWS has a 30 day limit, let's be conservative rather than
+		// hit the limit
+		maxAge = 60 * 60 * 24 * 28
+
+		// maxSkew allows for 5m of clock skew
+		maxSkew = 60 * 5
+	)
+
+	var (
+		content  = [traceIDLength]byte{}
+		epochNow = time.Now().Unix()
+		epoch    = int64(binary.BigEndian.Uint32(traceID[0:4]))
+		b        = [4]byte{}
+	)
+
+	// If AWS traceID originally came from AWS, no problem.  However, if oc generated
+	// the traceID, then the epoch may be outside the accepted AWS range of within the
+	// past 30 days.
+	//
+	// In that case, we use the current time as the epoch and accept that a new span
+	// may be created
+	if delta := epochNow - epoch; delta > maxAge || delta < -maxSkew {
+		epoch = epochNow
+	}
+
+	binary.BigEndian.PutUint32(b[0:4], uint32(epoch))
+
+	content[0] = '1'
+	content[1] = '-'
+	hex.Encode(content[2:10], b[0:4])
+	content[10] = '-'
+	hex.Encode(content[identifierOffset:], traceID[4:16]) // overwrite with identifier
+
+	return string(content[0:traceIDLength])
+}
+
+// convertToAmazonSpanID generates an Amazon spanID from a trace.SpanID - a 64-bit identifier
+// for the Segment, unique among segments in the same trace, in 16 hexadecimal digits.
+func convertToAmazonSpanID(v []byte) string {
+	if v == nil || reflect.DeepEqual(v, zeroSpanID) {
+		return ""
+	}
+	return hex.EncodeToString(v[0:8])
+}
+
+func timestampToFloatSeconds(ts *timestamp.Timestamp, startTs *timestamp.Timestamp) float64 {
+	var (
+		t time.Time
+	)
+	if ts == nil {
+		t = time.Now()
+	} else if startTs == nil {
+		t = time.Unix(ts.Seconds, int64(ts.Nanos))
+	} else {
+		t = time.Unix(startTs.Seconds, int64(ts.Nanos))
+	}
+	return float64(t.UnixNano()) / 1e9
+}
+
+func mergeAnnotations(dest map[string]interface{}, src map[string]string) {
+	for key, value := range src {
+		key = fixAnnotationKey(key)
+		dest[key] = value
+	}
+}
+
+func makeAnnotations(attributes map[string]string) (string, map[string]interface{}) {
+	var (
+		result = map[string]interface{}{}
+		user   string
+	)
+
+	delete(attributes, semconventions.AttributeComponent)
+	mergeAnnotations(result, attributes)
+
+	if len(result) == 0 {
+		return user, nil
+	}
+	return user, result
+}
+
+// fixSegmentName removes any invalid characters from the span name.  AWS X-Ray defines
+// the list of valid characters here:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+func fixSegmentName(name string) string {
```

Comment:
Add a test for SpanName with Invalid Characters. OR add a TODO to add the test later.

Reply:
Added test